File: languages/java/app/src/main/java/org/benchmarkplayground/benchmarkers/BenchmarkerManager.java
package org.benchmarkplayground.benchmarkers;

import java.security.InvalidKeyException;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

import org.benchmarkplayground.benchmarkers.impl.*;

public final class BenchmarkerManager {
    private static BenchmarkerManager BENCHMARKER_RUNNER = null;

    public static final BenchmarkerManager getBenchmarkerRunner() {
        // Guarantees that BenchmarkerRunner is only set once per application run
        if (BENCHMARKER_RUNNER == null) {
            BENCHMARKER_RUNNER = new BenchmarkerManager();
        }

        return BENCHMARKER_RUNNER;
    }

    private Map<String, Benchmarker> benchmarkers;

    private BenchmarkerManager() {
        benchmarkers = List.of(
            new QuickSortBenchmarker()
        ).stream()
        .collect(Collectors.toMap(Benchmarker::getOperationName, b -> b));
    }

    public Benchmarker getOperationBenchmarker(String operationName) throws InvalidKeyException {
        if (benchmarkers.containsKey(operationName)) {
            return benchmarkers.get(operationName);
        } else {
            String validNames = String.join(", ", getValidOperations());
            throw new InvalidKeyException(
                String.format("Operation name '%s' not found. Operation must be one of the following: %s", operationName, validNames)
            );
        }
    }

    private List<String> getValidOperations() {
        return benchmarkers.keySet().stream().sorted().toList();
    }
}

File: languages/c++/src/benchmarker/benchmarker_manager.cpp
#include <cstring>
#include <memory>
#include <vector>

#include "benchmarker.h"
#include "quick_sort_benchmarker.h"

namespace benchmarker {
    shared_ptr<Benchmarker> getBenchmarker(const char* operationName) {
        vector<Benchmarker*> benchmarkers = {
            new QuickSortBenchmarker()
        };
        for (Benchmarker* benchmarker : benchmarkers) {
            if (strcmp(benchmarker->getOperationName().c_str(), operationName) == 0) {
                return shared_ptr<Benchmarker>(benchmarker);
            }
        }

        return nullptr;
    }
}
